src/MiddlewareListener.php

```diff
@@ -47,15 +50,28 @@ public function onDispatch(MvcEvent $event)
         $application    = $event->getApplication();
         $response       = $application->getResponse();
         $serviceManager = $application->getServiceManager();
-        $middlewareName = is_string($middleware) ? $middleware : get_class($middleware);
 
-        if (is_string($middleware) && $serviceManager->has($middleware)) {
-            $middleware = $serviceManager->get($middleware);
-        }
-        if (! is_callable($middleware)) {
-            $return = $this->marshalMiddlewareNotCallable($application::ERROR_MIDDLEWARE_CANNOT_DISPATCH, $middlewareName, $event, $application);
-            $event->setResult($return);
-            return $return;
+        $psr7ResponsePrototype = Psr7Response::fromZend($response);
+        $pipe = new MiddlewarePipe();
+        $pipe->setResponsePrototype($psr7ResponsePrototype);
+
+        $middlewaresToBePiped = !is_array($middleware) ? [$middleware] : $middleware;
+
+        $middlewareName = 'noMiddlewarePiped';
+        $middlewareToBePiped = null;
+        foreach ($middlewaresToBePiped as $middlewareToBePiped) {
+            $middlewareName = is_string($middlewareToBePiped) ? $middlewareToBePiped : get_class($middlewareToBePiped);
+
+            if (is_string($middlewareToBePiped) && $serviceManager->has($middlewareToBePiped)) {
+                $middlewareToBePiped = $serviceManager->get($middlewareToBePiped);
+            }
+            if (! is_callable($middlewareToBePiped)) {
```

Comment:
If it's a string, it won't be a string any more. Also just following same logic that was already there...